File: cos/olympus/src/main/java/cos/olympus/game/Zone.java
package cos.olympus.game;

import cos.logging.Logger;
import cos.olympus.util.OpConsumer;
import cos.ops.out.CreatureHid;
import cos.ops.out.CreatureMoved;
import cos.ops.out.Metrics;
import cos.ops.out.ObjAppear;

public class Zone {

    private final static Logger logger = Logger.get(Zone.class);
    private final static int    radius = 8;

    private final World world;

    public Zone(World world) {
        this.world = world;
    }

    void onTick(Creature target, int tick, OpConsumer consumer) {
        //todo hardcode radius
        world.iterate(target.x, target.y, radius, (x, y) -> {
            var o = world.getObject(x, y);
            if (o != null && !target.zoneObjects.containsKey(o.id())) {
                target.zoneObjects.put(o.id(), o);
                consumer.add(new ObjAppear(o.id(), tick, target.id(), x, y, o.tile().id()));
            }

//            if (target.x == x && target.y == y) return; // avoid self-detection

            Creature cr;
            if (target.x == x && target.y == y) {
                cr = target;
            } else {
                cr = world.getCreature(x, y);
            }


            if (cr == null) {
                //disappear or /nothing
            } else {
                var ort = target.zoneCreatures.get(cr.id());
                if (ort == null || (ort.x() != cr.x || ort.y() != cr.y) || ort.speed() != cr.speed || ort.sight() != cr.sight) {
                    target.zoneCreatures.put(cr.id(), cr.orientation());
                    consumer.add(new CreatureMoved(1, tick, target.id(), cr.id(), x, y, cr.offset, cr.speed, cr.mv, cr.sight));
                }

                var met = target.zoneMetrics.get(cr.id());
                if (met == null || (met.life() != cr.life() || met.maxLife() != cr.metrics.maxLife())) {
                    var n = cr.copyMetrics();
                    target.zoneMetrics.put(cr.id(), n);
                    consumer.add(new Metrics(1, tick, target.id(), cr.id(), cr.level, n.life(), n.maxLife()));
                }
            }

        });

        target.zoneCreatures.values().removeIf(ort -> {
            if (ort.creatureId() == target.id()) return false;

            var cr = world.getCreature(ort.creatureId());
            if (cr == null || inNotFov(target, cr)) {
                consumer.add(new CreatureHid(1, tick, target.id(), ort.creatureId()));
                target.zoneMetrics.remove(ort.creatureId());
                return true;
            } else {
                return false;
            }
        });
    }

    static boolean inNotFov(Creature target, Creature o) {
        return Math.abs(target.x - o.x) > radius || Math.abs(target.y - o.y) > radius;
    }
}

File: cos/src/main/java/cos/olympus/game/Zone.kt
package cos.olympus.game

import cos.logging.Logger
import cos.olympus.util.OpConsumer
import cos.ops.out.Disappear
import cos.ops.out.Metrics
import cos.ops.out.Move
import cos.ops.out.Obj
import kotlin.math.abs

class Zone(private val world: World) {
    fun onTick(target: Actor, out: OpConsumer) {
        //todo hardcode radius
        world.iterateAround(target.x, target.y, VIEW_RADIUS) { x, y ->
            val obj = world.getObject(x, y)
            if (obj != null && !target.inZone(obj)) {
                target.addInZone(obj)
                out(Obj(obj.id, target.id, x, y, obj.tile.id))
            }

            val a = if (target.x == x && target.y == y) target else world.getActor(x, y)

            if (a == null) {
                //disappear or /nothing
            } else {
                val ort = target.zoneActors[a.id]
                if (ort == null || (ort.x != a.x || ort.y != a.y) || ort.speed != a.speed || ort.sight != a.sight) {
                    target.addInZone(a)
                    out(Move(1, target.id, a.id, x, y, a.offset, a.speed, a.mv, a.sight))
                }

                val met = target.zoneMetrics[a.id]
                if (a.metrics != met) {
                    val n = a.metrics.copy()
                    target.zoneMetrics[a.id] = n
                    out(Metrics(1, target.id, a.id, a.metrics.lvl, a.metrics.exp, n.life, n.maxLife))
                }
            }
        }

        target.zoneActors.values.removeIf { ort ->
            if (ort.actorId == target.id) return@removeIf false

            val a = world.getActor(ort.actorId)
            if (a == null || inNotFov(target, a)) {
                out(Disappear(1, target.id, ort.actorId))
                target.zoneMetrics.remove(ort.actorId)
                return@removeIf true
            } else {
                return@removeIf false
            }
        }
    }

    companion object {
        private val logger: Logger = Logger.get(Zone::class.java)
        private const val VIEW_RADIUS = 8

        fun inNotFov(target: Actor, o: Actor): Boolean {
            return abs((target.x - o.x).toDouble()) > VIEW_RADIUS || abs((target.y - o.y).toDouble()) > VIEW_RADIUS
        }
    }
}
